modules/tinymce/src/plugins/anchor/main/ts/core/Anchor.ts

```diff
@@ -5,37 +5,88 @@
+const isNamedAnchor = (elm: Node): elm is HTMLAnchorElement => isAnchor(elm) && !elm.getAttribute('href') && (!!elm.getAttribute('id') || !!elm.getAttribute('name'));
+const isEmptyNamedAnchor = (elm: Node): elm is HTMLAnchorElement => isNamedAnchor(elm) && !elm.firstChild;
+
+const removeEmptyNamedAnchorsInSelection = (editor: Editor) => {
+  const dom = editor.dom;
+  RangeUtils(dom).walk(editor.selection.getRng(), (nodes) => {
+    Arr.each(nodes, (node) => {
+      if (isEmptyNamedAnchor(node)) {
+        dom.remove(node, false);
+      }
+    });
+  });
+};
 
 const isValidId = (id: string) =>
   // Follows HTML4 rules: https://www.w3.org/TR/html401/types.html#type-id
   /^[A-Za-z][A-Za-z0-9\-:._]*$/.test(id);
 
-const getId = (editor: Editor) => {
-  const selectedNode = editor.selection.getNode();
-  return isNamedAnchor(editor, selectedNode) ? (selectedNode.getAttribute('id') || selectedNode.getAttribute('name')) : '';
-};
+const getNamedAnchor = (editor: Editor): Option<HTMLAnchorElement> => Option.from(editor.dom.getParent(editor.selection.getStart(), namedAnchorSelector) as HTMLAnchorElement);
 
-const insert = (editor: Editor, id: string) => {
-  const selectedNode = editor.selection.getNode();
+const getId = (editor: Editor) => getNamedAnchor(editor).map((anchor) => anchor.getAttribute('id') || anchor.getAttribute('name')).getOr('');
 
-  if (isNamedAnchor(editor, selectedNode)) {
-    selectedNode.removeAttribute('name');
-    selectedNode.id = id;
-    editor.undoManager.add();
-  } else {
-    editor.focus();
+const createAnchor = (editor: Editor, id: string): void => {
+  editor.focus();
+  if (!Settings.isHtmlInNamedAnchorAllowed(editor)) {
     editor.selection.collapse(true);
-    editor.insertContent(editor.dom.createHTML('a', {
-      id
-    }));
   }
+  if (editor.selection.isCollapsed()) {
+    editor.insertContent(editor.dom.createHTML('a', { id }));
+  } else {
+    editor.undoManager.transact(() => {
+      // Remove any empty named anchors in the selection as they cannot be removed by the formatter
+      removeEmptyNamedAnchorsInSelection(editor);
+      // Format is set up to truncate any partially selected named anchors so that they are not completely removed
+      editor.formatter.remove('namedAnchor', null, null, true);
+      // Insert new anchor using the formmater - will wrap selected content in anchor
+      editor.formatter.apply('namedAnchor', { value: id });
+      // Need to add visual classes to anchors if required
+      editor.addVisual();
+    });
+  }
+};
+
+const updateAnchor = (editor: Editor, id: string, anchorElement: HTMLAnchorElement): void => {
+  anchorElement.removeAttribute('name');
+  anchorElement.id = id;
+  editor.undoManager.add();
+};
+
+const insert = (editor: Editor, id: string) => getNamedAnchor(editor).fold(
+  () => createAnchor(editor, id),
+  (anchor) => updateAnchor(editor, id, anchor)
+);
+
+const registerFormat = (editor: Editor) => {
+  const namedAnchorFormat = {
```

Comment:
This should probably be part of core, as the `allow_html_in_named` anchor is a core functionality. You also mention this is similar to the way lists work and it also. registers the format in core, so that's further justification to add that there instead.

Reply:
Discussed via Slack and we'll keep it here for now. We can always move it later if needed.